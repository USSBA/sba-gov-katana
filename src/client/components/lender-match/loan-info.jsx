import React from 'react';
import {connect} from 'react-redux';
import {bindActionCreators} from 'redux';
import {CurrencyInput, TextArea, SelectBox} from '../helpers/form-helpers.jsx';
import MultiSelectBox from '../atoms/multiselect.jsx';
import * as LenderMatchActions from '../../actions/lender-match.js';
import * as LocationChangeActions from '../../actions/location-change.js';
import {getSelectBoxValidationState, getCurrencyValidationState, containsErrorOrNull, getTextAlphanumeicValidationState} from '../../services/page-validator-helpers.js'
import styles from './lender-match.scss';
import {includes, map} from "lodash";
import clientConfig from "../../services/config.js";

class LoanInfo extends React.Component {
  constructor(props) {
    super();
    let initialData = props.loanFields || {};
    this.state = {
      loanAmount: initialData.loanAmount || "",
      loanDescription: initialData.loanDescription || "",
      loanUsage: initialData.loanUsage || "",
      validStates: {
        loanAmount: null,
        loanDescription: null,
        loanUsage: null
      }
    };
  }

  componentDidMount() {
    this.validateFields(["loanAmount", "loanDescription", "loanUsage"]);
  }

  validateSingleField(validationFunction, name, defaultWhenNotSuccessful) {
    return validationFunction(name, this.state[name], defaultWhenNotSuccessful || null);
  }

  validateFields(fields, defaultWhenNotSuccessful) {
    let validStates = this.state.validStates;
    if (includes(fields, "loanAmount")) {
      validStates = Object.assign(validStates, this.validateSingleField(this.getLoanAmountValidationState, "loanAmount", defaultWhenNotSuccessful));
    }
    if (includes(fields, "loanDescription")) {
      validStates = Object.assign(validStates, this.validateSingleField(getSelectBoxValidationState, "loanDescription", defaultWhenNotSuccessful));
    }
    if (includes(fields, "loanUsage")) {
      validStates = Object.assign(validStates, this.validateSingleField(getTextAlphanumeicValidationState, "loanUsage", defaultWhenNotSuccessful));
    }
    this.setState({validStates: validStates});
  }

  isValidForm() {
    return !containsErrorOrNull(this.state.validStates);
  }

  getLoanAmountValidationState(name, value, defaultWhenNotSuccessful) {
    let validFormat = getCurrencyValidationState(name, value, defaultWhenNotSuccessful);
    let amountWithoutCommas = value.replace(/(\$|,)/g, "");
    let validAmount = Number(amountWithoutCommas) > 499;
    let valid = validFormat[name] === 'success' && validAmount ? "success" : defaultWhenNotSuccessful;
    let newState = {};
    newState[name] = valid;
    return newState;
  }

  handleSubmit(e) {
    e.preventDefault();
    this.props.actions.createLoan({loanAmount: this.state.loanAmount, loanDescription: this.state.loanDescription, loanUsage: this.state.loanUsage});
    this.props.locationActions.locationChange('/linc/form/additional', {
      action: "Continue Button Pushed",
      label: "/linc/form/loan"
    });
    this.loanForm.reset()
  }

  handleSelectChange(newValue) {
    this.setState({
      loanDescription: newValue
    }, () => this.validateFields(["loanDescription"]));
  }

  handleChange(e) {
    let name = e.target.name;
    let value = e.target.value;
    if (name && name === "loanDescription" && value && value.length > 250) {
      value = value.substring(0, 250);
    }
    if (name && name === "loanAmount") {
      let amountWithoutCommas = value.replace(/(\$|,)/g, "");
      if (amountWithoutCommas.length > 9) {
        value = this.state.loanAmount;
      }
    }
    let newState = {};
    newState[name] = value;
    this.setState(newState, () => this.validateFields([name]));
  }

  handleAmountBlur(value, callback) {
    let num = parseInt(value.replace(/(\$|,)/g, ""));
    let correctAmount = "";
    if (Number(num) || num === 0) {
      correctAmount = "$" + num.toLocaleString();
    }
    this.setState({
      loanAmount: correctAmount
    }, callback);
  }

  handleBlur(e) {
    let name = e.target.name;
    let value = e.target.value;
    if (name === "loanAmount") {
      this.handleAmountBlur(value, function() {
        this.validateFields([name], "error");
      });
    } else {
      this.validateFields([name], "error");
    }
  }

  render() {
    let loanDescriptionOptions = _.map([
      "Buying an Existing Business",
      "Developing a Product",
      "Hiring Employees/Staff",
      "Marketing/Advertising",
      "Opening a New Location ",
      "Other",
      "Participating in Trade Show",
      "Purchasing Equipment",
      "Purchasing Inventory",
      "Purchasing Property",
      "Refinancing/consolidating debt",
      "Remodeling an existing location",
      "Working Capital"
    ], (x) => {
      return {label: x, value: x};
    });

    return (
      <div>
        <form ref={(input) => this.loanForm = input} onSubmit={(e) => this.handleSubmit(e)}>
          <CurrencyInput errorText={clientConfig.messages.validation.invalidLoanAmount} label="How much funding do you need?" name="loanAmount" onChange={this.handleChange.bind(this)} onBlur={this.handleBlur.bind(this)} value={this.state.loanAmount} validationState={this.state.validStates.loanAmount} autoFocus/>
          <MultiSelectBox errorText={clientConfig.messages.validation.invalidLoanUsage} placeholder="- Select use of funds -" label="How will these funds be used?" name="loanDescription" onChange={this.handleSelectChange.bind(this)} validationState={this.state.validStates.loanDescription} value={this.state.loanDescription} options={loanDescriptionOptions} maxValues={3} onBlur={this.handleBlur.bind(this)}></MultiSelectBox>
          <TextArea errorText={clientConfig.messages.validation.invalidLoanDescription} label="Describe how you plan to use these funds" name="loanUsage" handleChange={this.handleChange.bind(this)} value={this.state.loanUsage} getValidationState={this.state.validStates.loanUsage} placeholder="I plan to purchase a larger oven to double the number of pizzas I can serve in an hour..." onBlur={this.handleBlur.bind(this)}/>
          <button className={styles.continueBtn} type="submit" disabled={!(this.isValidForm())}>
            CONTINUE
          </button>
        </form>
      </div>
    );
  };
}

function mapReduxStateToProps(reduxState) {
  return {loanFields: reduxState.lenderMatch.loanData};
}

function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators(LenderMatchActions, dispatch),
    locationActions: bindActionCreators(LocationChangeActions, dispatch)
  }
}

export default connect(mapReduxStateToProps, mapDispatchToProps)(LoanInfo);
