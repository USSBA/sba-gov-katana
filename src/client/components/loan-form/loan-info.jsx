import React from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { CurrencyInput, TextArea, SelectBox } from '../helpers/form-helpers.jsx'
import * as LoanActions from '../../actions/loan-form.js'
import { browserHistory } from 'react-router';
import { Col } from 'react-bootstrap';


class ContactInfoForm extends React.Component {
    constructor(){
        super();
        this.state ={
            loanFields: {
                loanAmount: ""
            }
        }
    }

    handleSubmit(e){
        e.preventDefault();
        this.props.actions.createLoan(this.state.loanFields);
        browserHistory.push('/form/business');
        this.loanForm.reset()
    }

    handleChange(e){
        let loanFields = {};
        loanFields[e.target.name] = e.target.value;
        this.setState({loanFields: {...this.state.loanFields, ...loanFields}});
        console.log(this.state.loanFields)
    }

    handleFormat(e){
        let loanFields = {};
        let num = parseInt(e.target.value.replace(/(\$|,)/g, ""));
        if(num && Number(e.target.value)) {
            loanFields[e.target.name] = "$" + num.toLocaleString() + ".00";
            this.setState({loanFields: {...this.state.loanFields, ...loanFields}});
        }
    }

    render() {
        return (
            <div>
                <form ref={(input) => this.loanForm = input} onSubmit={(e) => this.handleSubmit(e)}>
                    <CurrencyInput label="How much funding do you need?"
                                   name="loanAmount"
                                   handleChange={this.handleChange.bind(this)}
                                   handleFormat={this.handleFormat.bind(this)}
                                   value={this.state.loanFields.loanAmount}
                                   />

                    <SelectBox label="How will these funds be used?"
                               name="loanDescription"
                               handleChange={this.handleChange.bind(this)}
                               >
                        <option value="option 1">option 1</option>
                        <option value="option 2">option 2</option>
                    </SelectBox>

                    <TextArea label="Describe how these funds will be used?"
                              name="loanUsage"
                              handleChange={this.handleChange.bind(this)}
                              />

                        <button className="col-xs-2 col-xs-offset-5"
                                type="submit">
                                Next </button>

                </form>
            </div>
        );
    };
}

function mapReduxStateToProps(reduxState) {
    console.log(reduxState);
    return {};
}

function mapDispatchToProps(dispatch){
    return {
        actions: bindActionCreators(LoanActions, dispatch)
    }
}

export default connect(
    mapReduxStateToProps,
    mapDispatchToProps
)(ContactInfoForm);
