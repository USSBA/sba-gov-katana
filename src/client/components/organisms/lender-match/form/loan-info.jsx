import React from 'react'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import { includes, map, pickBy, keys } from 'lodash'
import { TextInput, TextArea, MultiSelectBox } from 'atoms'
import { FormPageButtons } from 'molecules'
import * as LenderMatchActions from '../../../../actions/lender-match.js'
import * as LocationChangeActions from '../../../../actions/navigation.js'
import {
  getSelectBoxValidationState,
  getCurrencyValidationState,
  containsErrorOrNull,
  getTextAlphanumeicValidationState
} from '../../../../services/form-validation-helpers.js'
import constants from '../../../../services/constants.js'
import { logEvent } from '../../../../services/analytics.js'

import styles from './lender-match.scss'

class LoanInfo extends React.Component {
  constructor(props) {
    super()
    let initialData = props.loanFields || {}
    this.state = {
      loanAmount: initialData.loanAmount || '',
      loanDescription: initialData.loanDescription || '',
      loanUsage: initialData.loanUsage || '',
      validStates: {
        loanAmount: null,
        loanDescription: null,
        loanUsage: null
      }
    }
  }

  componentDidMount() {
    this.validateFields(['loanAmount', 'loanDescription', 'loanUsage'])
    window.scrollTo(0, 0)
  }

  validateSingleField(validationFunction, name, defaultWhenNotSuccessful) {
    let validationState = validationFunction(
      name,
      this.state[name],
      defaultWhenNotSuccessful || null
    )
    if (validationState[name] === 'error') {
      logEvent({
        category: 'Lender Match Form',
        action: 'Error Event',
        label: name
      })
    }
    return validationState
  }

  validateFields(fields, defaultWhenNotSuccessful) {
    let validStates = this.state.validStates
    if (includes(fields, 'loanAmount')) {
      validStates = Object.assign(
        validStates,
        this.validateSingleField(
          this.getLoanAmountValidationState,
          'loanAmount',
          defaultWhenNotSuccessful
        )
      )
    }
    if (includes(fields, 'loanUsage')) {
      validStates = Object.assign(
        validStates,
        this.validateSingleField(
          getSelectBoxValidationState,
          'loanUsage',
          defaultWhenNotSuccessful
        )
      )
    }
    if (includes(fields, 'loanDescription')) {
      validStates = Object.assign(
        validStates,
        this.validateSingleField(
          getTextAlphanumeicValidationState,
          'loanDescription',
          defaultWhenNotSuccessful
        )
      )
    }
    this.setState({ validStates: validStates })
  }

  isValidForm() {
    return !containsErrorOrNull(this.state.validStates)
  }

  getLoanAmountValidationState(name, value, defaultWhenNotSuccessful) {
    let validFormat = getCurrencyValidationState(
      name,
      value,
      defaultWhenNotSuccessful
    )
    let amountWithoutCommas = value.replace(/(\$|,)/g, '')
    let validAmount = Number(amountWithoutCommas) > 499
    let valid =
      validFormat[name] === 'success' && validAmount
        ? 'success'
        : defaultWhenNotSuccessful
    let newState = {}
    newState[name] = valid
    return newState
  }

  handleSubmit(e) {
    e.preventDefault()
    this.props.actions.createLoan({
      loanAmount: this.state.loanAmount,
      loanDescription: this.state.loanDescription,
      loanUsage: this.state.loanUsage
    })
    this.props.locationActions.locationChange('/lendermatch/form/additional', {
      action: 'Continue Button Pushed',
      label: '/lendermatch/form/loan'
    })
    this.loanForm.reset()
  }

  handleSelectChange(newValue) {
    this.setState(
      {
        loanUsage: newValue
      },
      () => this.validateFields(['loanUsage'])
    )
  }

  handleChange(e) {
    let name = e.target.name
    let value = e.target.value
    if (name && name === 'loanDescription' && value && value.length > 250) {
      value = value.substring(0, 250)
    }
    if (name && name === 'loanAmount') {
      let amountWithoutCommas = value.replace(/(\$|,)/g, '')
      if (amountWithoutCommas.length > 9) {
        value = this.state.loanAmount
      }
    }
    let newState = {}
    newState[name] = value
    this.setState(newState, () => this.validateFields([name]))
  }

  handleAmountBlur(value, callback) {
    let num = parseInt(value.replace(/(\$|,)/g, ''))
    let correctAmount = ''
    if (Number(num) || num === 0) {
      correctAmount = '$' + num.toLocaleString()
    }
    this.setState(
      {
        loanAmount: correctAmount
      },
      callback
    )
  }

  handleBlur(e) {
    let name = e.target.name
    let value = e.target.value
    if (name === 'loanAmount') {
      this.handleAmountBlur(value, function() {
        this.validateFields([name], 'error')
      })
    } else {
      this.validateFields([name], 'error')
    }
  }

  handleFocus(nameOrEvent) {
    let name =
      nameOrEvent && nameOrEvent.target && nameOrEvent.target.name
        ? nameOrEvent.target.name
        : nameOrEvent
    logEvent({
      category: 'Lender Match Form',
      action: 'Focus Event',
      label: name
    })
  }

  render() {
    let loanUsageOptions = _.map(
      [
        'Purchasing property',
        'Buying equipment',
        'Remodeling/Expansion',
        'Refinancing debt',
        'Hiring employees',
        'Working capital',
        'Purchasing inventory',
        'Marketing',
        'Buy an existing business',
        'Developing a Product',
        'Opening a New Location',
        'Participating in Trade Show',
        'Other'
      ],
      x => {
        return { label: x, value: x }
      }
    )

    let id = 'lender-match-loan-info-form'
    return (
      <div>
        <form
          id={id}
          ref={input => (this.loanForm = input)}
          onSubmit={e => this.handleSubmit(e)}
        >
          <TextInput
            id={id + '-funding'}
            errorText={constants.messages.validation.invalidLoanAmount}
            label="How much funding do you need?"
            name="loanAmount"
            onChange={this.handleChange.bind(this)}
            onBlur={this.handleBlur.bind(this)}
            value={this.state.loanAmount}
            validationState={this.state.validStates.loanAmount}
            autoFocus
            onFocus={this.handleFocus.bind(this)}
          />
          <MultiSelectBox
            id={id + '-use'}
            errorText={constants.messages.validation.invalidLoanUsage}
            placeholder="- Select use of funds -"
            label="How will these funds be used?"
            name="loanUsage"
            onChange={this.handleSelectChange.bind(this)}
            validationState={this.state.validStates.loanUsage}
            value={this.state.loanUsage}
            options={loanUsageOptions}
            maxValues={3}
            onBlur={this.handleBlur.bind(this)}
            onFocus={this.handleFocus.bind(this)}
          />
          <TextArea
            id={id + '-describe'}
            errorText={constants.messages.validation.invalidLoanDescription}
            label="Describe how you plan to use these funds"
            name="loanDescription"
            onChange={this.handleChange.bind(this)}
            value={this.state.loanDescription}
            validationState={this.state.validStates.loanDescription}
            placeholder="I plan to purchase a larger oven to double the number of pizzas I can serve in an hour..."
            onBlur={this.handleBlur.bind(this)}
            onFocus={this.handleFocus.bind(this)}
          />
          <FormPageButtons
            parentId={id}
            backButtonHandler={this.props.locationActions.goBack}
            continueButtonHandler={this.handleSubmit.bind(this)}
            continueButtonDisabled={!this.isValidForm()}
          />
        </form>
      </div>
    )
  }
}

function mapReduxStateToProps(reduxState) {
  return { loanFields: reduxState.lenderMatch.loanData }
}

function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators(LenderMatchActions, dispatch),
    locationActions: bindActionCreators(LocationChangeActions, dispatch)
  }
}

export default connect(mapReduxStateToProps, mapDispatchToProps)(LoanInfo)
