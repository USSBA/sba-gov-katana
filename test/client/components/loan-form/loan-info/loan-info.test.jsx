import React from 'react';
import { LoanInfo } from '../../../../../src/client/components/loan-form/loan-info.jsx'
import renderer from 'react-test-renderer';


test('LoanInfo page renders inputs', () => {
    const component = renderer.create(
        <LoanInfo />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});
