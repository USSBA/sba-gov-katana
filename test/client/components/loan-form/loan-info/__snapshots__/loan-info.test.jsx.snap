exports[`test LoanInfo page renders inputs 1`] = `
<div
  className="col-xs-12 col-lg-6 col-lg-offset-3 formPanel panel panel-default"
  id={undefined}>
  <div
    className="panel-body">
    <div
      className="btn-toolbar"
      role="toolbar">
      <button
        className="btn btn-default btn-sm pull-left"
        onClick={[Function]}
        type="button">
        <span
          className="glyphicon glyphicon-chevron-left" />
         Back
      </button>
    </div>
    <h2
      className="col-md-4 col-md-offset-4 text-center">
      Loan Info
    </h2>
    <form
      onSubmit={[Function]}>
      <div
        className="col-sm-6 col-sm-offset-3 col-xs-12 col-xs-offset-0">
        <div
          className="input1 form-group has-feedback">
          <label
            className="controlLabel control-label"
            htmlFor={undefined}>
            How much funding do you need?
          </label>
          <input
            className="form-control"
            id={undefined}
            label="How much funding do you need?"
            name="loanAmount"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="" />
        </div>
      </div>
      <div
        className="col-sm-6 col-sm-offset-3 col-xs-12 col-xs-offset-0">
        <div
          className="form-group">
          <label
            className="controlLabel control-label"
            htmlFor={undefined}>
            How will these funds be used?
          </label>
          <select
            className="form-control"
            defaultValue=""
            id={undefined}
            name="loanDescription"
            onChange={[Function]}
            type={undefined}>
            <option
              disabled={true}
              value="">
              - Select use of funds -
            </option>
            <option
              value="option 1">
              option 1
            </option>
            <option
              value="option 2">
              option 2
            </option>
          </select>
        </div>
      </div>
      <div
        className="col-sm-6 col-sm-offset-3 col-xs-12 col-xs-offset-0">
        <div
          className="form-group">
          <label
            className="controlLabel control-label"
            htmlFor={undefined}>
            Describe how these funds will be used?
          </label>
          <textArea
            className="form-control"
            id={undefined}
            name="loanUsage"
            onChange={[Function]}
            placeholder="Include details such as this sample placeholder and this other example."
            type={undefined} />
        </div>
      </div>
      <button
        className="btn btn-default col-xs-2 col-xs-offset-5"
        type="submit">
        Continue 
      </button>
    </form>
  </div>
</div>
`;
